AWSAppSyncClient/AWSAppSyncServiceConfig.swift

```diff
@@ -0,0 +1,111 @@
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+//
+
+import Foundation
+
+/// Client-side configurations of an AWSAppSync service instance
+public protocol AWSAppSyncServiceConfigProviding {
+    /// The API endpoint
+    var endpoint: URL { get }
+
+    /// The AWS region of the API endpoint
+    var region: AWSRegionType { get }
+
+    /// The authentication method used to access the API. If this value is `.apiKey`, then the config must also provide a value
+    /// for `apiKey`
+    var authType: AWSAppSyncAuthType { get }
+
+    /// If `authType` is `.apiKey`, this value must provided
+    var apiKey: String? { get }
+}
+
+/// Client-side configurations of an AWSAppSync service instance
+public struct AWSAppSyncServiceConfig {
+    let endpoint: URL
+    let region: AWSRegionType
+    let authType: AWSAppSyncAuthType
+    let apiKey: String?
+
+    /// Reads configuration from `awsconfiguration.json` using the supplied key. If no key is supplied, reads configuration from
+    /// "Default".
+    ///
+    /// - Parameter key: The key from which to read the configuration values. Defaults to "Default"
+    /// - Throws: AWSAppSyncClientInfoError if the `awsconfiguration.json` file cannot be read, if the "AppSync" configuration
+    ///   is not present, or if the configuration is not valid.
+    public init(forKey key: String = "Default") throws {
+        let info = AWSInfo.default()
+
+        guard
+            let config = info.rootInfoDictionary[AWSConfigurationFile.Keys.root] as? [String: Any],
+            let configForKey = config[key] as? [String: Any]
+            else {
+                throw AWSAppSyncServiceConfigError.invalidConfigFile
+        }
+
+        guard
+            let apiURLString = configForKey[AWSConfigurationFile.Keys.apiURL] as? String,
+            let endpoint = URL(string: apiURLString)
+            else {
+                throw AWSAppSyncServiceConfigError.invalidAPIURL
+        }
+        self.endpoint = endpoint
+
+        guard
+            let regionString = configForKey[AWSConfigurationFile.Keys.region] as? String
+            else {
+                throw AWSAppSyncServiceConfigError.invalidRegion
+        }
+        self.region = regionString.aws_regionTypeValue()
+
+        guard
+            let authTypeRawValue = configForKey[AWSConfigurationFile.Keys.authMode] as? String,
+            let authType = AWSAppSyncAuthType(rawValue: authTypeRawValue)
+            else {
+                throw AWSAppSyncServiceConfigError.invalidAuthMode
+        }
+        self.authType = authType
+
+        apiKey = configForKey[AWSConfigurationFile.Keys.apiKey] as? String
+
+        guard authType != AWSAppSyncAuthType.apiKey || !apiKey.isEmpty else {
+            throw AWSAppSyncServiceConfigError.invalidAPIKey
+        }
+    }
+}
+
+public enum AWSAppSyncServiceConfigError: String, Error, LocalizedError {
```

Comment:
Just to consider. I usually extend classes separately from the main declaration, i.e.

```
// MARK: - LocalizedError

extension AWSAppSyncServiceConfigError: LocalizedError {
...
}

Reply:
Thanks for the note. I'm of two minds about this, and have done both ways (possibly in this very PR) by including the extension in the main body of the type and also doing a separate extension. As many others have noted, Swift doesn't allow you to define storage in an extension, so protocol conformance might be broken into multiple sections (base class for properties to support the protocol; an extension to define the behavior). However, that's probably the right thing to do anyway.

I'll consider this with the team and we'll make an official decision on the style we want to adopt for the project. For now, I'll move the conformance into a separate block as you suggest--I agree it keeps code more readable overall.

Reply:
Thanks for accepting note.

I think that this not only for readability. This approach also allows to throw away some obsolete or unnecessary code easily.

As far as I remember Apple Swift team also suggests this approach. But I cannot find link.

There’s some limitations, e.g. Codable / Hashable / Equatable auto synthesizing. You cannot extend class automatically with separate extension. But I personally prefer  separate extensions with some exceptions.

Reply:
Not Apple, but good team https://github.com/raywenderlich/swift-style-guide/blob/master/README.markdown#protocol-conformance